src/java.base/share/classes/java/lang/foreign/GroupLayout.java

```diff
@@ -73,4 +76,144 @@ public sealed interface GroupLayout extends MemoryLayout permits StructLayout, U
+     * <p>
+     * Unnamed elements in this group will be ignored.
+     * Unmatched elements (with respect to the name) in this group layout will be ignored.
+     * <p>
+     * The returned function will respect the byte orderings and alignment constraints of this
+     * group layout.
+     * <p>
+     * The returned Function may throw an {@link IllegalArgumentException} if it, for any reason, fails
+     * to extract a {@link Record}.  An example of such a failure is if the applied memory segment is too
+     * small for the layout at hand.
+     * <p>
+     * The example below shows how to extract an instance of a public {@code Point} record class
+     * from a {@link MemorySegment}:
+     * {@snippet lang = java:
+     *     MemorySegment segment = MemorySegment.ofArray(new int[]{3, 4});
+     *
+     *     public record Point(int x, int y){}
+     *
+     *     var pointLayout = MemoryLayout.structLayout(
+     *         JAVA_INT.withName("x"),
+     *         JAVA_INT.withName("y")
+     *     );
+     *
+     *     Function<MemorySegment, Point> pointExtractor = pointLayout.recordMapper(Point.class);
+     *
+     *     // Extracts a new Point from the provided MemorySegment
+     *     Point point = pointExtractor.apply(segment); // Point[x=3, y=4]
+     * }
+     * <p>
+     * Boxing, widening and narrowing must be explicitly handled by user code.  In the following example, the above
+     * {@code Point} (using primitive {@code int x} and {@code int y} coordinates) are explicitly mapped to
+     * a narrowed point type (instead using primitive {@code byte x} and {@code byte y} coordinates):
+     * <p>
+     * {@snippet lang = java:
+     *     public record NarrowedPoint(byte x, byte y) {
+     *
+     *         static NarrowedPoint fromPoint(Point p) {
+     *             return new NarrowedPoint((byte) p.x, (byte) p.y);
+     *         }
+     *     }
+     *
+     *     Function<MemorySegment, NarrowedPoint> narrowedPointExtractor =
+     *             pointLayout.recordMapper(Point.class)
+     *                     .andThen(NarrowedPoint::fromPoint);
+     *
+     *     // Extracts a new NarrowedPoint from the provided MemorySegment
+     *     NarrowedPoint narrowedPoint = narrowedPointExtractor.apply(segment); // NarrowedPoint[x=3, y=4]
+     * }
+     *
+     * @param <R> record type
+     * @param type the type (Class) of the record
+     * @throws IllegalArgumentException if the provided record {@code type} is the class {@link Record} or contains
+     *                                  components for which there are no exact mapping (of names and types) in
+     *                                  this group layout or if the provided {@code type} is not public or
+     *                                  if the method is otherwise unable to create a record mapper as specified above.
+     * @since 21
+     */
+    <R extends Record> Function<MemorySegment, R> recordMapper(Class<R> type);
```

Comment:
I feel like the documentation is a bit _too_ formal, and will go over some people's heads. In general, I suggest using plain English as the main structure, and then weaving in the formal definitions instead.

I think by re-ordering some of the information, you can end up with simpler/smaller paragraphs. I think there are 3 main steps:
1. The `recordMapper` function itself finds, for each record component, a corresponding member layout with the same name in this group layout. There are restrictions on the layout type of each corresponding member layout based on the type of the record component (e.g. a java primitive maps to a ValueLayout, MemorySegment maps to an AddressLayout, etc.)
2. An extraction function is determined based on the component/layout type combination (and then have a list with the process for each type combination). In this part you can lean on the information defined in 1, and use language like "the corresponding member layout"/"the corresponding record component", rather than `c_a` and `e_b`.
3. The canonical constructor of the record type is called with the result of the extraction functions.

I think 1 & 2 are too mixed up right now in the big formal definition of the extraction functions.

Reply:
I am happy to simplify. What do you think @mcimadamore ?